### ISSUE TITLE
Improve separation of concerns: model, optimization, parameter update rules

### ISSUE THREAD

**Issue by aarangop** [aarangop/rust-ml#1]
Currently params are passed around using hashmaps, and bein cloned. Implement a different mechanism to allow the optimizer to update parameters. The optimizer doesn't necessarily need access to model parameters, it just needs a way to signal that they should be updated. This opens the door to implementing a parameter update mechanism that's independent of model and optimizer, in order to allow for different update methods apart from standard gradient descent, like RMSProps, or ADAM, which require additional hyperparamters and distinct update rules.

**Comment by aarangop** [comment]
Implemented model-related traits (OptimizableModel, BaseModel), Optimizer, and Optimizer state.

Each trait defines methods separating concern, broadly:
- Model traits: Define model functionality, and allow accessing and mutating model paraters.
- Optimizer: Orchestrates optimization, for now, only gradient descent has been implemented
- OptimizerState: Apply parameter update rules.

This trait system was made thinking about cases where update rules might be different but independent from the optimization algorithm, for instance, we could choose to train a model using mini batch gradient descent, but update the parameters with momentum. The momentum update is managed by OptimizerState, while the Optimizer creates the mini batches batches and coordinates individual training steps, while OptimizerState is ultimately responsible for updating parameters' values.
